// carbon-components-react/lib/components/Dialog/Dialog-story.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DialogExample = exports.Default = exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _FocusScope = require("../FocusScope");

var _Dialog = require("../Dialog");

var _useId = require("../../internal/useId");

var _Portal = require("../Portal");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var _default = {
  title: 'Experimental/unstable_Dialog',
  includeStories: []
};
exports.default = _default;

var Default = function Default() {
  function DemoComponent() {
    var _React$useState = React.useState(false),
        _React$useState2 = _slicedToArray(_React$useState, 2),
        open = _React$useState2[0],
        setOpen = _React$useState2[1];

    var ref = React.useRef(null);
    return /*#__PURE__*/React.createElement("div", {
      style: {
        border: '1px solid black',
        background: 'rgba(0, 0, 0, 0.1)',
        padding: '1rem'
      }
    }, /*#__PURE__*/React.createElement("button", {
      type: "button",
      onClick: function onClick() {
        setOpen(true);
      }
    }, "Open"), open ? /*#__PURE__*/React.createElement(_FocusScope.FocusScope, null, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("p", null, "Elit hic at labore culpa itaque fugiat. Consequuntur iure autem autem officiis dolores facilis nulla earum! Neque quia nemo sequi assumenda ratione officia Voluptate beatae eligendi placeat nemo laborum, ratione."), /*#__PURE__*/React.createElement(DemoComponent, null), /*#__PURE__*/React.createElement("button", {
      ref: ref,
      type: "button",
      onClick: function onClick() {
        setOpen(false);
      }
    }, "Close"))) : null);
  }

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(DemoComponent, null), /*#__PURE__*/React.createElement("button", {
    type: "button"
  }, "Hello"));
};

exports.Default = Default;

var DialogExample = function DialogExample() {
  function Example() {
    var _React$useState3 = React.useState(false),
        _React$useState4 = _slicedToArray(_React$useState3, 2),
        open = _React$useState4[0],
        setOpen = _React$useState4[1];

    var id = (0, _useId.useId)();
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("button", {
      type: "button"
    }, "First")), /*#__PURE__*/React.createElement("button", {
      type: "button",
      onClick: function onClick() {
        setOpen(true);
      }
    }, "Open"), open ? /*#__PURE__*/React.createElement(_Portal.Portal, {
      style: {
        position: 'fixed',
        top: 0,
        right: 0,
        bottom: 0,
        left: 0,
        zIndex: 9999
      }
    }, /*#__PURE__*/React.createElement(FullPage, null), /*#__PURE__*/React.createElement(_Dialog.Dialog, {
      "aria-labelledby": id,
      onDismiss: function onDismiss() {
        setOpen(false);
      },
      style: {
        position: 'relative',
        zIndex: 9999,
        padding: '1rem',
        background: 'white'
      }
    }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("span", {
      id: id
    }, "Hello")), /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Example, null)), /*#__PURE__*/React.createElement("button", {
      type: "button",
      onClick: function onClick() {
        setOpen(false);
      }
    }, "Close"))) : null, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("button", {
      type: "button"
    }, "Last")));
  }

  return /*#__PURE__*/React.createElement(Example, null);
};

exports.DialogExample = DialogExample;
var FullPage = /*#__PURE__*/React.forwardRef(function FullPage(props, ref) {
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    style: {
      position: 'absolute',
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
      transform: 'translateZ(0)',
      background: 'rgba(0, 0, 0, 0.5)'
    }
  }, props));
});
